src/lmdb/database.cpp

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2014-2018, The Monero Project
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#include "database.h"
+#include "lmdb/error.h"
+#include "lmdb/util.h"
+
+#ifdef _WIN32
+namespace
+{
+    constexpr const mdb_mode_t open_flags = 0;
+}
+#else
+#include <sys/stat.h>
+
+namespace
+{
+    constexpr const mdb_mode_t open_flags = (S_IRUSR | S_IWUSR);
+}
+#endif
+
+namespace lmdb
+{
+    namespace
+    {
+        void acquire_context(context& ctx) noexcept
+        {
+            while (ctx.lock.test_and_set());
```

Comment:
Why not a mutex ? Faster if no contention ?

Reply:
Yes, the thinking is that the contention will be low and therefore this will be faster. This is basically a direct copy from the current lmdb blockchain code.

Boost has a spinlock implementation in detail and which contains some basic yield code if there is a bit of contention. The only problem is that being in the detail namespace means no documentation and no guarantee of API stability (yes, even less than normal!).

I also read through the GNU pthread mutex implementation. The current implementation has a user space locking mechanism before during a system call to block the thread. Its _fairly_ quick, but there is some extra book-keeping work when acquiring or releasing the lock to handle thread suspension. Resuming suspended threads is faster than the boost yield code - the blocking thread "knows" which threads are waiting for it to complete.

I'll look into leveraging the boost yield code. A high majority of the time this is being held for an atomic increment and then released, and in rare cases this blocks during a LMDB map resize. The boost code will help with backoff in that rare case, and adds proper pauses that intel recommends, etc.